### PR TITLE
Force integral depth value for layer-wise PF clusters

### PR DESCRIPTION
**Pull request by marksan87**
#### PR description:

This addresses #35923 by setting the PFCluster depth to the PFRecHit integer depth when all rechits have the same depth value. The PFCluster depth takes on non-integer values for multidepth clusters, so this parameter must remain a double. 

#### PR validation:

Passes all matrix tests in `runTheMatrix.py -l limited -i all --ibeos`. As shown below, HBHE PFCluster depths no longer contain rounding errors. 

HBHE_PFCluster_depth_comparison

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not a backport

@hatakeyamak @bendavid @rappoccio @laurenhay @silviodonato

### CODE REVIEW

File: RecoParticleFlow/PFClusterProducer/plugins/PFMultiDepthClusterizer.cc [hunk 1]
```diff
@@ -175,10 +175,10 @@ std::vector<PFMultiDepthClusterizer::ClusterLink> PFMultiDepthClusterizer::link(
       const reco::PFCluster& cluster1 = clusters[i];
       const reco::PFCluster& cluster2 = clusters[j];
 
-      auto dz = (cluster2.depth() - cluster1.depth());
+      auto dz = ((int)cluster2.depth() - (int)cluster1.depth());
 
       //Do not link at the same layer and only link inside out!
-      if (dz < 0.0f || std::abs(dz) < 0.2f)
+      if (dz < 0.0 || std::abs(dz) < 0.2)
```
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/plugins/PFMultiDepthClusterizer.cc:181]: does the second condition matter after the conversion to int?
```suggestion
      if (dz <= 0.0)
```

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by silviodonato** [comment]
I suggest also to remove `if (fabs(link1.dZ() - link2.dZ()) < 0.2)` in https://github.com/cms-sw/cmssw/pull/36088/files#diff-21080d7b6fa996faacee9a4a1923257c50d72796ec14a7a1d1b8fd9ec51a0b16R224
as this condition is always true.
I can make a separated PR if you prefer.

**Comment by hatakeyamak** [comment]
@kdlong @juska

**Comment by slava77** [comment]
> I suggest also to remove `if (fabs(link1.dZ() - link2.dZ()) < 0.2)` in https://github.com/cms-sw/cmssw/pull/36088/files#diff-21080d7b6fa996faacee9a4a1923257c50d72796ec14a7a1d1b8fd9ec51a0b16R224 as this condition is always true. I can make a separated PR if you prefer.

please point to the line of code, for me the link above leads just to the top of the page in https://github.com/cms-sw/cmssw/pull/36088/files

**Comment by slava77** [comment]
> > I suggest also to remove `if (fabs(link1.dZ() - link2.dZ()) < 0.2)` in https://github.com/cms-sw/cmssw/pull/36088/files#diff-21080d7b6fa996faacee9a4a1923257c50d72796ec14a7a1d1b8fd9ec51a0b16R224 as this condition is always true. I can make a separated PR if you prefer.
> 
> please point to the line of code, for me the link above leads just to the top of the page in https://github.com/cms-sw/cmssw/pull/36088/files

nm, I found it in 
https://github.com/cms-sw/cmssw/blob/f132666fced55d63bdb7bff780293b3a5326e230/RecoParticleFlow/PFClusterProducer/plugins/PFMultiDepthClusterizer.cc#L224

**Comment by VinInn** [comment]
should dz in link become an integer?
I advice to review all logic based on "dz"

**Comment by slava77** [comment]
@marksan87 @hatakeyamak 
please clarify if the PFCluster parameters are expected (the kinematics of `particleFlowClusterHCAL` changes according to the comparisons). Only depth is provided in the PR description. Some plot for energy consistency or smth related would be useful to confirm that this update did not uncover something unexpected.

**Comment by hatakeyamak** [comment]
> @marksan87 @hatakeyamak please clarify if the PFCluster parameters are expected (the kinematics of `particleFlowClusterHCAL` changes according to the comparisons). Only depth is provided in the PR description. Some plot for energy consistency or something related would be useful to confirm that this update did not uncover something unexpected.

Plan is to run a standard on-the-fly PF validation to make sure we don't find anything unusual in any of PF candidates of PF jets/MET like quantities. Does this sound a reasonable check?

**Comment by slava77** [comment]
> @slava77 Our plan is to run a standard on-the-fly PF validation to make sure we don't find anything unusual in any of PF candidates of PF jets/MET like quantities. Does this sound a reasonable check?

should be OK for me.
Thank you.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by marksan87** [comment]
> > > @slava77 Our plan is to run a standard on-the-fly PF validation to make sure we don't find anything unusual in any of PF candidates of PF jets/MET like quantities. Does this sound a reasonable check?
> > 
> > 
> > should be OK for me. Thank you.
> 
> @hatakeyamak @laurenhay @marksan87 please clarify on the status of this check. Thank you.

@slava77 please see the following slides for validation checks comparing multidepth HCAL PFClusters, PFCandidates, and PFJets on CPU and GPU with these changes. The CPU/GPU agreement for PFJets is greatly improved. 
[PFCluster Depth CPU_GPU comparisons.pdf](https://github.com/cms-sw/cmssw/files/7549002/PFCluster.Depth.CPU_GPU.comparisons.pdf)

**Review by slava77 — COMMENTED** [review]
(somehow I never pushed a submit button for the case with a cast)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
Just to be sure, considering the ongoing discussion in  #35923:
is it reasonable to consider this PR good to go and that if any further changes in depth treatment, it will come from a follow up PR?

**Comment by marksan87** [comment]
> Just to be sure, considering the ongoing discussion in #35923: is it reasonable to consider this PR good to go and that if any further changes in depth treatment, it will come from a follow up PR?

I think so. This PR should be considered a bug fix. Any further changes needed to address #35923 can be included in a separate PR.

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/36088	7429731 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show small differences starting from `particleFlowClusterHCAL` and propagating downstream
- higher stat comparisons shown in https://github.com/cms-sw/cmssw/pull/36088#issuecomment-970565500 show that the physics difference is apparently small and the main feature update is in the significant improvement between the CPU and GPU results

**Comment by perrotta** [comment]
+1

**Comment by silviodonato** [comment]
For the record, Ganesh from TSG confirm that this PR solved the problem of the GPU/CPU differences in PF in HCAL
(see https://docs.google.com/spreadsheets/d/1MOAzxYSRKRjqDvkQKUd6a9SAT0mNO--_3Zc3w9JCtQ0/edit#gid=853920917 )

**Pull request merged**